### ISSUE TITLE
Heartbeats fails to send after upgrade to 2.3.0

### ISSUE THREAD

**Issue by YC** [muety/wakapi#342]
**Describe the bug**
POST heartbeats returns "bad request" after upgrade to 2.3.0. 
Returns to normal after downgrade to 2.2.5. 

**System information**
* Wakapi version: Docker 2.3.0
* Database; SQLite

Logs:
```
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/cmd/run.go:108","func":"Run","level":"debug","message":"command: heartbeat","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/cmd/heartbeat/heartbeat.go:86","func":"SendHeartbeats","level":"debug","message":"params: api params: (api key: '<hidden>cc33', api url: 'https://<redacted>/time/api', backoff at: '2022-03-19T14:49:53+11:00', backoff retries: 1, hostname: 'tc', plugin: 'vscode/1.65.2 vscode-wakatime/18.0.7', timeout: 2m0s, disable ssl verify: false, proxy url: '', ssl cert filepath: ''), heartbeat params: (category: 'coding', cursor position: '1', entity: '/home/steve/.wakatime.log', entity type: 'file', num extra heartbeats: 0, is write: true, language: '', line number: '1', lines in file: '1', time: 1647661830.38055, filter params: (exclude: '[]', exclude unknown project: false, include: '[]', include only with project file: false), project params: (alternate: 'steve', disable submodule: '[]', map patterns: '[]', override: ''), sanitize params: (hide branch names: '[]', hide project folder: false, hide file names: '[]', hide project names: '[]', project path override: '')), offline params: (disabled: false, queue file: '', num sync max: 1000), status bar params: (hide categories: false)","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/heartbeat/format.go:25","func":"1","level":"debug","message":"execute heartbeat filepath formatting","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/filter/filter.go:28","func":"1","level":"debug","message":"execute heartbeat filtering","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/heartbeat/entity_modify.go:15","func":"1","level":"debug","message":"execute heartbeat entity modifier","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/remote/remote.go:47","func":"1","level":"debug","message":"execute remote file detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/filestats/filestats.go:29","func":"1","level":"debug","message":"execute filestats detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/language/language.go:19","func":"1","level":"debug","message":"execute language detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/language/language.go:40","func":"1","level":"debug","message":"failed to detect language on file entity \"/home/steve/.wakatime.log\": could not detect the language of file \"/home/steve/.wakatime.log\"","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/deps/deps.go:38","func":"1","level":"debug","message":"execute dependency detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/project.go:69","func":"1","level":"debug","message":"execute project detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/file.go:22","func":"Detect","level":"debug","message":"execute file project detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/map.go:29","func":"Detect","level":"debug","message":"execute map project detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/git.go:23","func":"Detect","level":"debug","message":"execute git project detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/mercurial.go:19","func":"Detect","level":"debug","message":"execute mercurial project detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/subversion.go:21","func":"Detect","level":"debug","message":"execute subversion project detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/tfvc.go:18","func":"Detect","level":"debug","message":"execute tfvc project detection","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/project/filter.go:22","func":"1","level":"debug","message":"execute project filtering","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/heartbeat/sanitize.go:33","func":"1","level":"debug","message":"execute heartbeat sanitization","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/offline/offline.go:52","func":"1","level":"debug","message":"execute offline queue with file /home/steve/.wakatime.bdb","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/backoff/backoff.go:40","func":"1","level":"debug","message":"execute heartbeat backoff algorithm","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/backoff/backoff.go:78","func":"shouldBackoff","level":"debug","message":"exponential backoff tried 1 times since Mar 19 14:49:53, will retry at Mar 19 14:50:23","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/api/heartbeat.go:26","func":"SendHeartbeats","level":"debug","message":"sending 1 heartbeat(s) to api at https://<redacted>/time/api/users/current/heartbeats.bulk","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/api/heartbeat.go:33","func":"SendHeartbeats","level":"debug","message":"heartbeats: [{\"branch\":\"\",\"category\":\"coding\",\"cursorpos\":1,\"dependencies\":null,\"entity\":\"/home/steve/.wakatime.log\",\"type\":\"file\",\"is_write\":true,\"language\":null,\"lineno\":1,\"lines\":1,\"project\":\"\",\"time\":1647661830.3805456,\"user_agent\":\"wakatime/v1.38.0 (linux-5.16.14-arch1-1-unknown) go1.17.7 vscode/1.65.2 vscode-wakatime/18.0.7\"}]","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/backoff/backoff.go:48","func":"1","level":"debug","message":"incrementing backoff due to error","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/offline/offline.go:62","func":"1","level":"debug","message":"pushing 1 heartbeat(s) to queue due to error","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/cmd/run.go:246","func":"runCmd","level":"error","message":"failed to run command: sending heartbeat(s) later due to bad request: bad request at \"https://<redacted>/time/api/users/current/heartbeats.bulk\"","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/pkg/api/diagnostic.go:28","func":"SendDiagnostics","level":"debug","message":"sending diagnostic data to api at https://<redacted>/time/api/plugins/errors","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
{"caller":"/home/runner/work/wakatime-cli/wakatime-cli/cmd/run.go:252","func":"runCmd","level":"warning","message":"failed to send diagnostics: failed to send diagnostics to the API: invalid response status from \"https://<redacted>/time/api/plugins/errors\". got: 401, want: 201. body: \"401 unauthorized\"","now":"2022-03-19T14:50:30+11:00","version":"v1.38.0"}
```

**Comment by muety** [comment]
Can you show an example of a failing heartbeat, please?

**Comment by muety** [comment]
The only thing that have changed in 2.3.0 to that regard is that the endpoint now doesn't accept any more heartbeats that are older than a configurable age (6 months by default, see https://github.com/muety/wakapi/blob/e7e5254673b45557f9e54c6372d4979089f244b9/config.default.yml#L19). But I assume you're not trying to send heartbeats that old?

**Comment by YC** [comment]
```
[19/Mar/2022:06:49:37 +0000] "POST /time/api/users/current/heartbeats.bulk HTTP/2.0" 400 138 "" "wakatime/v1.38.0 (linux-5.16.14-arch1-1-unknown) go1.17.7 vscode/1.65.2 vscode-wakatime/18.0.7" "[{\"branch\":\"\",\"category\":\"coding\",\"cursorpos\":1,\"dependencies\":null,\"entity\":\"/home/steve/.wakatime.cfg\",\"type\":\"file\",\"is_write\":null,\"language\":\"INI\",\"lineno\":1,\"lines\":13,\"project\":\"\",\"time\":1647672611.8321865,\"user_agent\":\"wakatime/v1.38.0 (linux-5.16.14-arch1-1-unknown) go1.17.7 vscode/1.65.2 vscode-wakatime/18.0.7\"}]"
[19/Mar/2022:06:49:37 +0000] "POST /time/api/plugins/errors HTTP/2.0" 401 134 "" "wakatime/v1.38.0 (linux-5.16.14-arch1-1-unknown) go1.17.7 vscode/1.65.2 vscode-wakatime/18.0.7" ""
[19/Mar/2022:06:49:40 +0000] "POST /time/api/users/current/heartbeats.bulk HTTP/2.0" 400 138 "" "wakatime/v1.38.0 (linux-5.16.14-arch1-1-unknown) go1.17.7 vscode/1.65.2 vscode-wakatime/18.0.7" "[{\"branch\":\"\",\"category\":\"coding\",\"cursorpos\":1,\"dependencies\":null,\"entity\":\"/home/steve/.wakatime.cfg\",\"type\":\"file\",\"is_write\":true,\"language\":\"INI\",\"lineno\":8,\"lines\":9,\"project\":\"\",\"time\":1647672614.7582898,\"user_agent\":\"wakatime/v1.38.0 (linux-5.16.14-arch1-1-unknown) go1.17.7 vscode/1.65.2 vscode-wakatime/18.0.7\"}]"
[19/Mar/2022:06:49:40 +0000] "POST /time/api/plugins/errors HTTP/2.0" 401 134 "" "wakatime/v1.38.0 (linux-5.16.14-arch1-1-unknown) go1.17.7 vscode/1.65.2 vscode-wakatime/18.0.7" ""
```

**Comment by muety** [comment]
Might be a time zone issue. Can you please run this little script both on your local machine and inside the Docker container?

```go
package main

import (
	"fmt"
	"time"
)

func main() {
	_, off := time.Now().Zone()
	fmt.Println(off)
}
```

**Comment by YC** [comment]
Locally: 39600
Remotely (I used alpine:3 image, added golang): 0

**Comment by muety** [comment]
I'll keep you posted.

**Comment by muety** [comment]
If you try to post one of the above heartbeats manually (using curl or Postman), do you get the same error?

**Comment by muety** [comment]
The only reason for this error, which I can think of right now, is, that your local machine's clock and the server clock that runs Wakapi (are you even running it on a server?) are a bit off. Since last release, Wakapi also doesn't allow for heartbeats from the future, so if your server clock is a couple of seconds behind, it will error out. I push a fix to tolerate 1 hour differences. But not sure, if that's actually your issue.

**Comment by YC** [comment]
I used one of the 2 from above and it worked. 
Yes, I am running it on a remote server. 
My laptop does also have a tendency to drift. 
`19 Mar 20:38:00 ntpdate[76615]: step time server 129.250.35.251 offset -34.130315 sec`

**Comment by muety** [comment]
Alright, then that is very likely the reason! Will be fixed with 2.3.1.

**Comment by YC** [comment]
Great, thanks!
